src/types.jl

```diff
@@ -244,20 +274,37 @@ const ArbTypes = Union{
     AcbRefMatrix,
 }
 
-Base.setindex!(x::Union{Mag,Arf,Arb,ArbRef,Acb,AcbRef}, z::Number) = set!(x, z)
+Base.setindex!(x::Union{Mag,MagRef,Arf,ArfRef,Arb,ArbRef,Acb,AcbRef}, z::Number) =
+    set!(x, z)
+Base.setindex!(x::Union{Mag,MagRef}, z::Ptr{mag_struct}) = set!(x, z)
+Base.setindex!(x::Union{Arf,ArfRef}, z::Ptr{arf_struct}) = set!(x, z)
 Base.setindex!(x::Union{Arb,ArbRef}, z::Ptr{arb_struct}) = set!(x, z)
 Base.setindex!(x::Union{Acb,AcbRef}, z::Ptr{acb_struct}) = set!(x, z)
 
-Base.Float64(x::Mag) = get(x)
-function Base.Float64(
-    x::Union{Arf,Ptr{arf_struct}};
-    rnd::Union{arb_rnd,RoundingMode} = RoundNearest,
-)
+function midref(x::Union{Arb,ArbRef}, prec = _precision(x))
+    mid_ptr = ccall(@libarb(arb_mid_ptr), Ptr{arf_struct}, (Ref{arb_struct},), x)
+    ArfRef(mid_ptr, prec, parentstruct(x))
+end
+function radref(x::Union{Arb,ArbRef})
+    rad_ptr = ccall(@libarb(arb_rad_ptr), Ptr{mag_struct}, (Ref{arb_struct},), x)
+    MagRef(rad_ptr, parentstruct(x))
+end
+
+function realref(z::Union{Acb,AcbRef}; prec = _precision(z))
+    real_ptr = ccall(@libarb(acb_real_ptr), Ptr{arb_struct}, (Ref{acb_struct},), z)
+    ArbRef(real_ptr, prec, parentstruct(z))
+end
+function imagref(z::Union{Acb,AcbRef}; prec = _precision(z))
+    real_ptr = ccall(@libarb(acb_imag_ptr), Ptr{arb_struct}, (Ref{acb_struct},), z)
+    ArbRef(real_ptr, prec, parentstruct(z))
+end
+
+Base.Float64(x::MagLike) = get(x)
+function Base.Float64(x::ArfLike; rnd::Union{arb_rnd,RoundingMode} = RoundNearest)
     ccall(@libarb(arf_get_d), Cdouble, (Ref{arf_struct}, arb_rnd), x, rnd)
 end
-function Base.Int(
-    x::Union{Arf,Ptr{arf_struct}};
-    rnd::Union{arb_rnd,RoundingMode} = RoundNearest,
-)
+function Base.Int(x::ArfLike; rnd::Union{arb_rnd,RoundingMode} = RoundNearest)
     ccall(@libarb(arf_get_si), Clong, (Ref{arf_struct}, arb_rnd), x, rnd)
 end
+Base.Float64(x::ArbLike) = Float64(midref(x))
+Base.ComplexF64(z::AcbLike) = Complex(Float64(realref(z)), Float64(imagref(z)))
```

Comment:
These might make more sense to have in a `conversion.jl` file. But that can be fixed at a later point.

Reply:
Make sense, but this probably best to group together with maybe some other regrouping.